### DISCUSSION TITLE
StringBuilder: char constructor

### DISCUSSION THREAD

**Discussion by eduherminio** [dotnet/runtime#46052] [General] [answered]
After (painfully) realizing that certain `StringBuilder` constructor accepts a `char`, but not the one I was expecting to, I'm now wondering whether adding the following constructor to `StringBuilder` has ever been considered:

```csharp
public StringBuilder(char? value)
```

This is a simplified version of the code that brought me here:

```csharp
var str = "ABC";
var sb = new StringBuilder(str.Last());

Console.WriteLine($"-->{sb.ToString()}<--");
```

Thanks!

**Comment by PathogenDavid** [comment]
I feel like this isn't a common enough situation to justify adding a new API. I'm assuming you *need* a `StringBuilder` in this situation (if not, the `StringBuilder` is adding a ton of overhead for a single character string and you should just use `str.Last().ToString()`.)

If you need to allocate a single-character `StringBuilder` and you want to do so efficiently, you can explicitly specify the initial capacity to be 1. IE:

```cs
var str = "ABC";

var sb = new StringBuilder(1);
sb.Append(str.Last());
```

**Comment by eduherminio** [comment]
I would have said that initializing a `StringBuilder` with a `char` would be a relatively common situation, but if no one proposed it before, I guess you're right and it's not 🤷🏼‍♂️.

I still think that it would be great to avoid the current behavior difference beween `new StringBuilder("a")` and `new StringBuilder('a')`.

PD: as stated, that was a simplified version of the code, the `StringBuilder` instance was initialized with a specific `char`, but appended large strings chunks afterwards.

**Reply by PathogenDavid** [reply]
> I still think that it would be great to avoid the current behavior difference beween `new StringBuilder("a")` and `new StringBuilder('a')`.

~~I would not expect there to be a significant performance difference if the single-character string passed in is constant. If the difference is impactful enough to matter to you, you probably shouldn't even be using `StringBuilder` in the first place.~~

Oops, I thought this was about a perceived performance difference. I didn't originally read this correctly:

> After (painfully) realizing that certain `StringBuilder` constructor accepts a `char`, but not the one I was expecting to

What unfortunately happened was the `char` was implicitly cast to `int`. So it called the initial-capacity constructor.

Unfortunately that means adding a new constructor would break existing code. (Even if that code is probably incorrect.)

This was proposed before as https://github.com/dotnet/runtime/issues/20848 and was rejected. You can watch the design review for their reasoning here: https://youtu.be/JdaQ2RP2tqM?t=6187

I think this could be something that could make sense as an analyzer. Possibly even as one of the analyzers in [`Microsoft.CodeAnalysis.NetAnalyzers`](https://github.com/dotnet/roslyn-analyzers/blob/028d9e0c1ad9e5e80ad503d35a804bd62210ecbb/src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.md).

**Reply by eduherminio** [reply]
I somehow missed that issue 😓 thanks @PathogenDavid!

**Reply by PathogenDavid** [reply]
No problem! I've found it helps to restrict the search to the titles of issues. I found it using [`is:issue stringbuilder char in:title `](https://github.com/dotnet/runtime/issues?q=is%3Aissue+stringbuilder+char+in%3Atitle)

**Comment by danmoseley** [comment]
Feel free to open issue proposing an analyzer.

**Reply by danmoseley** [reply]
BTW I wished to mark @PathogenDavid's 2nd response above as the "Answer" but it seems one cannot tag responses to comments as answers.

**Reply by PathogenDavid** [reply]
Yeah, I probably should've posted that part separately. I extracted the answer half of that comment into a proper answer if you still want to mark it for the sake of posterity.

**Reply by danmoseley** [reply]
done!

**Reply by eduherminio** [reply]
Awesome, I opened https://github.com/dotnet/runtime/issues/46103 for it, don't know why GH isn't linking it here though.
Thanks!

**Comment by PathogenDavid** [comment, marked as answer]
> After (painfully) realizing that certain `StringBuilder` constructor accepts a `char`, but not the one I was expecting to

What unfortunately happened was the `char` was implicitly cast to `int`. So it called the initial-capacity constructor.

Unfortunately that means adding a new constructor would break existing code. (Even if that code is probably incorrect.)

This was proposed before as https://github.com/dotnet/runtime/issues/20848 and was rejected. You can watch the design review for their reasoning here: https://youtu.be/JdaQ2RP2tqM?t=6187

I think this could be something that could make sense as an analyzer. Possibly even as one of the analyzers in [`Microsoft.CodeAnalysis.NetAnalyzers`](https://github.com/dotnet/roslyn-analyzers/blob/028d9e0c1ad9e5e80ad503d35a804bd62210ecbb/src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.md).
